### ISSUE TITLE
 readers/mutation_readers: compacting_reader forgets about injected partition-end

### ISSUE THREAD

**Issue by denesb** [scylladb/scylladb#11608]
Currently injecting a partition-end doesn't update `_last_uncompacted_kind`, which will allow for a subsequent `next_partition()` call to trigger injecting a partition-end, leading to an invalid mutation fragment stream (partition-end after partition-end).

**Comment by avikivity** [comment]
What scenario can cause this to happen (before the fix)? What's the impact (beyond having an invalid mutation fragment stream)?

**Comment by denesb** [comment]
Two consecutive `next_partition()` calls. The compacting reader is now used in cache and compaction (and was used during memtable flush). I don't know what the impact is, but the fix is so small we can backport out of caution.
Once I added the validator, the issue would reproduce every single time a test started up a node (used `cql_test_env`).

**Comment by denesb** [comment]
The extra `partition_end` fragments are emitted to the compactor and is ignored by the reader. This mechanism (injecting partition-end into the stream on `next_partition()`) was added to this reader because the compactor it uses expects each partition to end properly with a `partition_end`, even if it is not consumed completely. The consumer of the reader however doesn't want a `partition_end` fragment when calling `next_partition()`. So I added a mechanism which in injects this `partition_end` into the compactor's input stream but ignores it in the compactor's output stream. All this is to say that this bug is completely benign AFAIK and only started being a problem after a validator was added to the compactor, which now suddenly didn't like consecutive `partition_end` fragments.
